pex/third_party/__init__.py

```diff
@@ -0,0 +1,392 @@
+# Copyright 2018 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import absolute_import
+
+import contextlib
+import importlib
+import os
+import re
+import sys
+
+import warnings
+import zipfile
+from collections import namedtuple
+from contextlib import contextmanager
+
+
+# NB: All pex imports are performed lazily to play well with the un-imports performed by both the
+# PEX runtime when it demotes the bootstrap code and any pex modules that uninstalled
+# VendorImporters un-import.
+def _tracer():
+  from pex.tracer import TRACER
+
+  return TRACER
+
+
+class _Loader(namedtuple('_Loader', ['module_name', 'vendor_module_name'])):
+
+  # The PEP-302 loader API.
+  # See: https://www.python.org/dev/peps/pep-0302/#specification-part-1-the-importer-protocol
+  def load_module(self, fullname):
+    assert fullname in (self.module_name, self.vendor_module_name), (
+      '{} got an unexpected module {}'.format(self, fullname)
+    )
+    vendored_module = importlib.import_module(self.vendor_module_name)
+    sys.modules[fullname] = vendored_module
+    _tracer().log('{} imported via {}'.format(fullname, self), V=9)
+    return vendored_module
+
+  def unload(self):
+    for mod in (self.module_name, self.vendor_module_name):
+      if mod in sys.modules:
+        sys.modules.pop(mod)
+        _tracer().log('un-imported {}'.format(mod), V=9)
+
+        submod_prefix = mod + '.'
+        for submod in sorted(m for m in sys.modules.keys() if m.startswith(submod_prefix)):
+          sys.modules.pop(submod)
+          _tracer().log('un-imported {}'.format(submod), V=9)
+
+
+class _Importable(namedtuple('_Importable', ['module', 'is_pkg', 'path', 'prefix'])):
+  _exposed = False  # noqa: We want instance variable access defaulting to cls here.
+
+  def expose(self):
+    self._exposed = True
+    importlib.import_module(self.module)
+    _tracer().log('Exposed {}'.format(self), V=3)
+
+  def loader_for(self, fullname):
+    _, _, target = fullname.partition(self.prefix + '.')
```

Comment:
Would be helpful to use `_my_ignored_var` style if you know what the first two are. Not a big deal.

Reply:
I couldn't find a way to make this clear and uncluttered at the same time - switched to a test-then-chomp.

Reply:
Actually I think `_, _, thing_i_need = foo.partition(...)` is very idiomatic. In general, and particularly with `partition`.

Reply:
I left this at zag instead of zigging back to the original. A recurring theme is definitely that Eric wants more explicit names - long flag names where possible in bash scripts and subprocess calls, and favoring `_unused_but_named` over `_`. These two preferences are at odds with the pex codebase generally and I haven't pushed really hard on underlying motivations. I will note that it could be argued in the `_` case that the very point of `_` is "doesn't matter" and so these might be read as "don't read me": ie focus on the other code. As such, switching from `_` to `_distracting_you_a_little_by_naming_my_intent_if_actually_used` is ... just distracting.

Reply:
I prefer the `_` usage for precisely the reason you mention: giving something a name implies to the reader that it matters, when it doesn't. It's just more cognitive load for no reason.

Reply:
Agreed that I have through a couple reviews been pushing for more explicitness. Much of this derives from my lack of experience with PEX, making it sometimes challenging for me to figure out what certain CLI flags represent or an abbreviated variable stands for. Another motivation is personal style, that to me it seems more readable. I recognize, however, that with open source, codebase conventions should trump personal preferences.

You both raise solid arguments for `_` over `_unused_but_named`. For how I reason through code, the latter style helps me to infer what the function is doing. However, the former style is completely valid and there are strong arguments for why it is preferred. Especially because PEX uses the convention of `_`, I think you're both right that we should stick to that style, or refactor it as John did with the two uses I flagged.

The one point I will continue to advocate for is long CLI flags. Even if that is not the convention with PEX, I always think the readability of long flags trumps saving a few characters for short flags. When using the command line directly, I use short flags, but when writing it in a file saved to VCS, I advocate for long flags.

P.S. Congratulations on merging 🎉

Reply:
Yeah, I think those are different cases. The flags are meaningful. Using the long version helps the reader. But the unused variables are not, and naming them distracts the reader. 

The general rule of naming and comments I like to follow is "assume the reader knows the language and the std libraries as well as you, but doesn't know what you're trying to do with them".